lib/code-builder/src/lib.rs

```diff
@@ -0,0 +1,181 @@
+
+#![feature(specialization)]
+
+use basegl_prelude::*;
+use std::fmt::Write;
+
+
+// ===============
+// === Builder ===
+// ===============
+
+/// The main code building structure. It is like a string-builder but also knows about the current
+/// indentation and contains information if it is needed to add a space when adding a new symbol.
+/// For example, you can add `foo` and add `bar` symbols and the resulting code will contain
+/// `foo bar` instead of (obviously wrong) `foobar`.
+#[derive(Clone,Debug)]
+pub struct CodeBuilder {
+    pub spaces_in_indent : usize,
+    pub indent           : usize,
+    pub spaced           : bool,
+    pub buffer           : String
+}
+
+impl CodeBuilder {
+    /// Interface to the underlying `Writer` implementation. Please note that this function does not
+    /// take into account the indentation nor spacing, so other smart code constructors are
+    /// recommended in most cases.
+    pub fn write<S:Str>(&mut self, s:S) {
+        self.write_str(s.as_ref()).unwrap();
+    }
+
+    /// Increase the indentation of the code. Will be used by the next `newline` call.
+    pub fn inc_indent(&mut self) {
+        self.indent += 1;
+    }
+
+    /// Decrease the indentation of the code. Will be used by the next `newline` call.
+    pub fn dec_indent(&mut self) {
+        self.indent -= 1;
+    }
+
+    /// Create a new line and insert appropriate indentation.
+    pub fn newline(&mut self) {
+        let space_count = self.spaces_in_indent * self.indent;
+        self.write(format!("\n{}"," ".repeat(space_count)));
+        self.spaced = true;
+    }
+
+    /// Inserts a terminator symbol `;`.
+    pub fn terminator(&mut self) {
+        self.write(";");
+        self.spaced = false;
+    }
+
+    /// Adds a new element to the builder. The element can be any form of string, slice, or any
+    /// other object which implements the `HasCodeRepr` trait.
+    pub fn add<T>(&mut self, t:T)
+        where Self: AddToBuilder<T> {
+        self.add_to_builder(t)
+    }
+
+    /// Adds a new element and assumes it is spaced (there is no need to add a space before
+    /// inserting the next element).
+    pub fn add_spaced<T>(&mut self, t:T)
+        where Self: AddToBuilder<T> {
+        self.add(t);
+        self.spaced = true;
+    }
+
+    /// Specialization of the `add` method for strings.
+    fn add_str<S:Str>(&mut self, s:S) {
+        if !self.spaced {
+            self.write(" ");
+        }
+        self.spaced = false;
+        self.write(s);
+    }
+}
+
+// === AddToBuilder ===
+
+/// A helper trait for handling polymorphic input to the `add` method. It allows adding all kind of
+/// strings (including slices) and also any objects that implement the `Printer` trait.
+pub trait AddToBuilder<T> {
+    fn add_to_builder(&mut self, t:T);
+}
+
+impl<T: HasCodeRepr> AddToBuilder<&T> for CodeBuilder {
+    default fn add_to_builder(&mut self, t:&T) {
+        t.build(self)
+    }
+}
```

Comment:
Can we? I don't know a way to do it.

Reply:
It'll look like this way:
In Builder impl:
```rust
    pub fn add<T>(&mut self, t:&T) -> &mut Self
        where T: HasCodeRepr<T> {
        t.build(self);
        self
    }
```
And implement HasCodeRepr for String:
```rust
impl HasCodeRepr for String {
    fn build(&self, builder:&mut CodeBuilder) {
        builder.add_str(self);
    }
}
```
The only downside is that `add` function must take a reference. But this does not change much, because eventually from all Strings we're getting as_str/as_ref

Reply:
Oh, yep, it will work this way. If we have problem with the reference, then we can change `HasCodeRepr`'s `build` to one which takes ownership of self.